### ISSUE TITLE
Termination problems when querying Linked Data profile

### ISSUE THREAD

**Issue by RubenVerborgh** [comunica/comunica#144]
#### Issue type:

- :bug: Bug

____
#### Case 1
```bash
$ comunica-sparql http://ruben.verborgh.org/profile/ 'SELECT * { ?s ?p ?o } LIMIT 1'
```
gives a correct result but does not terminate. (Note the `http:`.)

#### Case 2
```bash
$ comunica-sparql https://ruben.verborgh.org/profile/ 'SELECT * { ?s ?p ?o } LIMIT 1'
```
gives a `socket hang up`. (Note the `https:`.)

#### Case 3
The `https:` socket issue disappears when removing the `LIMIT 1`, but it still does not terminate then.

____
#### Environment:

| software            | version
| ------------------- | -------
| Comunica Init Actor | 1.1.0
| node                | v10.0.0
| npm                 | 6.1.0
| yarn                | _Yarn is unavailable_
| Operating System    | darwin (Darwin 13.4.0)

#### Log file:

For second case:
```
events.js:167
      throw er; // Unhandled 'error' event
      ^

Error: socket hang up
    at createHangUpError (_http_client.js:312:15)
    at TLSSocket.socketCloseListener (_http_client.js:344:23)
    at TLSSocket.emit (events.js:187:15)
    at _handle.close (net.js:598:12)
    at TCP.done [as _onclose] (_tls_wrap.js:387:7)
Emitted 'error' event at:
    at ClientRequest.eventHandlers.(anonymous function) (/Users/ruben/.nvm/versions/node/v10.0.0/lib/node_modules/@comunica/actor-init-sparql/node_modules/follow-redirects/index.js:16:24)
    at ClientRequest.emit (events.js:182:13)
    at TLSSocket.socketCloseListener (_http_client.js:344:9)
    at TLSSocket.emit (events.js:187:15)
    at _handle.close (net.js:598:12)
    at TCP.done [as _onclose] (_tls_wrap.js:387:7)
```

**Comment by rubensworks** [comment]
Ok, the problem seems to be that comunica limits the stream using asynciterator's `range()`.
This causes the asynciterator to stop consuming elements from the stream, which behind the scenes is piped from the HTTP response. The application however hangs because the HTTP response has not been consumed completely (or closed/destroyed) in this case.

@RubenVerborgh It looks like asynciterator does not emit the `close` event. And the question is if it should, in order to make situations like this detectable, so that the HTTP response can be properly closed in this case.

A quick, but possibly less optimal, fix would be to make the buffered iterator more eager, and always consume the whole stream.

**Comment by RubenVerborgh** [comment]
> Ok, the problem seems to be that comunica limits the stream using asynciterator's range().

Do you really mean [range](https://rubenverborgh.github.io/AsyncIterator/docs/AsyncIterator.html#.range), which creates a new iterator of integers, or (for instance) [take](https://rubenverborgh.github.io/AsyncIterator/docs/AsyncIterator.html#take), which sets a limit?

> @RubenVerborgh It looks like asynciterator does not emit the close event.

It does not, only has readable/data/end/error.

Note that the [Node documentation](https://nodejs.org/api/stream.html#stream_event_close_1) says:

> Not all Readable streams will emit the 'close' event.

> And the question is if it should, in order to make situations like this detectable, so that the HTTP response can be properly closed in this case.

Could you sketch how this would work this in some more detail? I don't fully follow.

> A quick, but possibly less optimal, fix would be to make the buffered iterator more eager, and always consume the whole stream.

That is [the approach taken in Client.js](https://github.com/LinkedDataFragments/Client.js/blob/v2.0.5/lib/util/HttpClient.js#L68). Just FYI, not necessarily advocating it.

**Comment by rubensworks** [comment]
> Do you really mean range, which creates a new iterator of integers, or (for instance) take, which sets a limit?

I mean [the non-static range](https://rubenverborgh.github.io/AsyncIterator/docs/AsyncIterator.html#range), which does a combined limit and offset.

> Could you sketch how this would work this in some more detail? I don't fully follow.

Ok, let's consider the following.
```
let quadStream = httpResponse.pipe(parserStream);
... some code to transform quadStream in a bindingsStream (asynciterator)...
let newBindingsStream = bindingsStream.range(offset, offset + limit);
```
Ideally, when all elements from `newBindingsStream` have been consumed, the `bindingsStream` should be closed (either by asynciterator, or by comunica, not sure about that yet). This close event would cascade back through the whole piping chain until `httpResponse` (I don't think that Node does this by itself, I'd have to check), so that the HTTP response stream is properly closed.

> That is the approach taken in Client.js. Just FYI, not necessarily advocating it.

That might be a good solution for now. But at some point in the future, I'd like to investigate very large HTTP responses that act as streams, where the client should be able to interrupt them at any point.

**Comment by RubenVerborgh** [comment]
> I mean the non-static range, which does a combined limit and offset.

That moment when you forget the methods in your own library 😄 

> Ideally, when all elements from newBindingsStream have been consumed, the bindingsStream should be closed (either by asynciterator, or by comunica, not sure about that yet).

So what you actually want then is a `closeSourceWhenDone` property in the constructor of a `TransformIterator`?

> This close event would cascade back through the whole piping chain

That's the part where I don't follow with the events, since events bubble down, not up. So `bindingsStream` would never listen to `newBindingsStream`.

Note that this is where `AsyncIterator` deviates from Node streams, since these would listen to each other (notably the `writable` event).

>     That is the approach taken in Client.js. Just FYI, not necessarily advocating it.
> 
> That might be a good solution for now.

Maybe, but a full fix would also be nice.

**Comment by rubensworks** [comment]
> So what you actually want then is a closeSourceWhenDone property in the constructor of a TransformIterator?

That would be a useful addition :-)
Alternatively, we could do the same as is needed for native Node streams (see below)

> That's the part where I don't follow with the events, since events bubble down, not up.

Indeed, that would require us to add backwards-closing handlers ourselves. But I'm hesitant to do that, as that would be needed on each `pipe` call on non-asynciterators. I'm still hoping to find a cleaner solution.

**Comment by RubenVerborgh** [comment]
How many non-asynciterators are we still dealing with? I'd assume that the first thing you do with such an object is make it an AsyncIterator with https://rubenverborgh.github.io/AsyncIterator/docs/AsyncIterator.html#.wrap .

**Comment by rubensworks** [comment]
> How many non-asynciterators are we still dealing with?

Quite a few actually. Most quad streams are pure Node streams, so that we are compatible with RDFJS at that level. From the moment that we enter the bindings-world, everything is an asynciterator.

**Comment by RubenVerborgh** [comment]
Right. Would `closeSourceWhenDone` solve our problems in both cases? If so, feel free to start a PR on AsyncIterator.

**Comment by rubensworks** [comment]
> Right. Would closeSourceWhenDone solve our problems in both cases?

It should directly solve case 1. The others might require some other changes, not sure yet, but that won't have anything to do with the streams.

> If so, feel free to start a PR on AsyncIterator.

👌

**Comment by RubenVerborgh** [comment]
Actually thinking that `closeSourceWhenDone` should be `true` by default.
Currently, the docs of https://rubenverborgh.github.io/AsyncIterator/docs/AsyncIterator.html#range state that

> The current iterator may not be read anymore until the returned iterator ends.

but we could make that stronger such that

> After this operation, only read the returned iterator instead of the current one.

as is the case with other operations.

**Comment by rubensworks** [comment]
Makes sense, as the documentation does not say anything about the state of the untransformed iterator after the transformed iterator ends, so reusing that is probably not very useful for developers at the moment.

**Comment by rubensworks** [comment]
Note to self: everything in the non-AsyncIterator world works as intended. All streams are push-based, and no streams will be left unconsumed that can cause the application to hang.

The only remaining problem is when handling AsyncIterators. Following up in https://github.com/RubenVerborgh/AsyncIterator/issues/6

**Comment by rubensworks** [comment]
So in the end no code changes were needed to comunica to fix this. See https://github.com/RubenVerborgh/AsyncIterator/issues/6 and https://github.com/olalonde/follow-redirects/pull/88
Once these PRs are merged and the packages update, then we can update their versions here.

**Comment by RubenVerborgh** [comment]
follow-redirects has a new release, and the latest develop in AsyncIterator should have the right fix. Can you confirm with AsyncIterator? Then I'll release 2.0.

**Comment by rubensworks** [comment]
When updating to follow-redirects 1.5.1 and AsyncIterator on the develop branch, both the http and https issues as described in this issue are resolved.

**Comment by RubenVerborgh** [comment]
Thanks! Published `asynciterator 2.0.0`.
